### ISSUE TITLE
Form still get submitted even if having syncErrors

### ISSUE THREAD

**Issue by chaintng** [erikras/redux-form#1356]
I use `redux-form/immutable` version v6.0.0-rc.3

However, after i implement it, the form still able to submit even there is a syncError object underneath the form store.

When submit the form, everything goes well, the error of each inptu are shown
but it won't stop the submit process.

I track down the library and it look like it didn't have `this.submitPromise`
![image](https://cloud.githubusercontent.com/assets/2979072/16894470/509b9c7c-4b81-11e6-8ad0-6564cfa2f2d5.png)

I have no idea how to debug more, if you want any other information, please let me know.

in CustomerForm.js

```
class CustomerForm extends React.Component { // eslint-disable-line react/prefer-stateless-function
  render() {
    const { handleSubmit, submitting } = this.props;
    console.log(handleSubmit);
    return (
      <form className={styles.customerForm} onSubmit={handleSubmit}>
        <div>
          <label>Full Name: </label>
          <Field component={renderField} type="text" name="fullName" />
        </div>
        <div>
          <label>Phone Number: </label>
          <Field component={renderField} type="text" name="phoneNumber" />
        </div>
        <button type="submit" disabled={submitting}>Submit</button>
      </form>
    );
  }
}

CustomerForm.propTypes = {
  onSubmit: React.PropTypes.any,
  handleSubmit: React.PropTypes.any,
  submitting: React.PropTypes.any,
  customerDetail: React.PropTypes.object,
};

CustomerForm = reduxForm({
  form: 'customerDetail',
  validate,
})(CustomerForm);
```

and in its parent

```
const showResults = values =>
  new Promise(resolve => {
    setTimeout(() => {  // simulate server latency
      window.alert(`You submitted:\n\n${JSON.stringify(values, null, 2)}`)
      resolve()
    }, 500)
  })


export class Register extends React.Component { // eslint-disable-line react/prefer-stateless-function
  render() {
    return (
      <div className={styles.register}>
        <h1>Register</h1>
        <CustomerForm
          onSubmit={showResults}
          customerDetail={this.props.customerDetail}
        />
      </div>
    );
  }
}
```

**Comment by igorbarbashin** [comment]
What does your `validate` function look like?

**Comment by igorbarbashin** [comment]
If you use `redux-form/immutable` then I believe you need to convert data to js in the validate function.
Like so:
```javascript
  const {
    fullName,
    phoneNumber,
  } = values.toJS();
```

Though I'm not sure this is the source of the issue.

**Comment by chaintan17** [comment]
I check `values.fullName` and it's already a string.
The comparison is work right when user input in input box.
But look like when submitting the form, it didn't go in to validation function once more...
I set breakpoint there but it didn't stop.

**Comment by igorbarbashin** [comment]
Hm, looks like you're doing everything right.
The only thing I might try is to put your validate function the same file with the form. And make sure that you don't have js errors when loading the page. Cause if there's errors, the form is submitting no matter what with a regular "GET" method. 

I don't quite get what else could be wrong with your form.

Something to do with `validate` function

**Comment by mattjstar** [comment]
Maybe this might help, but I ran into a similar issue. 

We have a newsletter form in the footer of our React app. We initially server render all of our pages with `ReactDOM.renderToString`, but then use react-router to go between our pages on the client. Weirdly, we noticed that the `validate` function was only working properly on those subsequent client rendered page loads. I'm currently debugging the issue now, but we're going with only rendering our forms on the client until we have a fix.

**Comment by mattjstar** [comment]
Actually, scratch that. I discovered the issue on our end: https://github.com/erikras/redux-form/blob/v6.0.0-rc.3/src/reduxForm.js#L129

Why do we also check for `!_plain2.default.deepEqual(syncErrors, nextSyncErrors)`? If I remove that part of the if statement everything starts working as expected.

Since we server render our form initially, the `'redux-form/INITIALIZE`' and `'redux-form/UPDATE_SYNC_ERRORS'` actions are already called. When we get to the `updateSyncErrorsIfNeeded` function inside of the `Form` HOC, `syncErrors` and `nextSyncErrors` are equal, so `updateSyncErrors` isn't called. Unfortunately, there's a race condition when we hydrate our client as `'redux-form/INITIALIZE'` gets called again and wipes out our errors but `this.props.syncErrors` is pointing to the server rendered `syncErrors` object still in memory.

I could open up a separate issue if we think this is different.

**Comment by chaintan17** [comment]
@mattjstar unfortunately, it's look different issue to my case. I render my form all in react.
I have one more clue, compare to an example here, when user click `Submit` button, redux show new action called `redux-form/TOUCH` and `redux-form/SET_SUBMIT_FAILED`
But in my form, it didn't have both action...

**Comment by chaintan17** [comment]
I try changing the example app to `redux-form/immutable` and it has the same issue
(chainging this line: https://github.com/erikras/redux-form/blob/v6.0.0-rc.3/examples/syncValidation/src/SyncValidationForm.js#L2)

After changes, the validation for textbox changes, are work perfectly.
But with submit button, it can still go through.

**Comment by at0g** [comment]
I'm experiencing the same issue as @mattjstar regarding the universal render.
In my case I'm not using immutable, but sync validation fails to run on an untouched form IF the form was rendered server-side. Client side render = works as expected.

**Comment by tyrsius** [comment]
Im on [6.0.0-rc.4] and I am expericencing this issue as well. Errors are even showing on the form, but if you press the submit button the submit action is called. Short of manually re-calling validation in submit, I have no way to stop the user from saving invalid forms.

**Comment by tyrsius** [comment]
Im on [6.0.0-rc.4] and I am expericencing this issue as well. Errors are even showing on the form, but if you press the submit button the submit action is called. Short of manually re-calling validation in submit, I have no way to stop the user from saving invalid forms.

@erikras  this seems like a serious enough issue to hold up the release

**Comment by tyrsius** [comment]
Actually, it's important to note that the `onSubmit` function is only called with a `FieldArray` has errors, like this one

![image](https://cloud.githubusercontent.com/assets/1187432/17608306/7f9df286-5fe0-11e6-9c3c-1d27d2bdbeca.png)

An error on a normal field *does* stop `onSubmit` from being called, but nested `_error` props will not stop `onSubmit`. The [FieldArray example](http://redux-form.com/6.0.0-rc.1/examples/fieldArrays/) shows that this is indeed the correct response.

**Comment by erikras** [comment]
This is because the `_error` key for an array is not compatible with using `redux-form/immutable`.

There is some discussion about a workaround on #1310.

**Comment by tyrsius** [comment]
I'm not using redux-form/immutable.

**Comment by erikras** [comment]
@tyrsius Do you have a `<Field name="grantTypes" component={Whatever}/>`?

**Comment by tyrsius** [comment]
Yes

**Comment by KaboomFox** [comment]
Happens to me if the form is pristine and I click submit. Sync Validation is not called or returns valid. Async validation is called and `syncValidationPasses` returns true when it should be false.

**Comment by marekjalovec** [comment]
this stopped happening to me after upgrading to 6.0.1

**Comment by davelnewton** [comment]
I'm seeing this behavior in 6.1.0. I'm not using `immutable`.

My `<Field>`s are in a separate file from my form, wrapped in a small component, and also use the example `renderField` function. My `onSubmit` handler is in the same file as the form, outside of the component that renders the form (but was also tried as a class property). 

If I touch the field validation happens properly. If I just hit submit without having edited the field submission happens. A log statement in the `validate` method never fires, so validation isn't being called on submission.

If I pull back everything into the component that renders the form the behavior is the same, e.g., it wasn't because I went overboard refactoring. I see no meaningful difference between my code and the `syncValidation` example. This behavior occurs from 6.0.1-6.1.0.

**Comment by davelnewton** [comment]
Followup on my own report:

I have resolved the issue, although I'm not sure which change I made actually did it. I had at least two issues that caused this: I was not passing the submit handler as a property to the component containing the form, I was only putting it on the form. 

I realized this when the `values` being passed into the submit handler were the event, not the form values. I'm assuming (but haven't checked) that whatever we pass in gets wrapped around/under a form's normal submit handler. I'll need to restructure my code to take this into account, but I think I'm good for now--apologies for the noise.

**Comment by vam** [comment]
I can confirm this behavior on 6.1.1. It happens when a form has only a non field error (i.e. `_error` war set) while all fields are valid.

I beleive that a part of the problem is the following line of of the CHANGE event handler in the reducer:
```js
result = deleteInWithCleanUp(result, 'syncError');
```
The syncError property will be deleted __every time__ a CHANGE action is dispatched. Also, the action UPDATE_SYNC_ERRORS, which could set the `syncError` again, is not triggered as needed.

**Comment by Amurmurmur** [comment]
@vam Has this been resolved?
I currently use 
`import { Field, reduxForm } from 'redux-form/immutable'`
And form still gets submitted even with error shown :/

**Comment by vam** [comment]
@Amurmurmur We've moved our project to 6.2.1 and I can't confirm that problem in that version.

**Comment by sachinmour** [comment]
I still have this problem.. in the latest version

**Comment by erikras** [comment]
@sachinmour Can you recreate it in [a sandbox](https://github.com/erikras/redux-form#-code-sandboxes-)?

**Comment by MartinDawson** [comment]
@sachinmour @erikras @Amurmurmur 

I was having the same issue. The form must be passed an `onSubmit` handler which redux-form passes the values to and then returns a `handleSubmit` function which must be passed as a prop to form, e.g. `onSubmit={handleSubmit}`.

I expected validation to work without having to do this.

**Comment by danielrob** [comment]
Closing as this is resolved for some, and details are thin from others. If this is not resolved, please provide reproduction steps here, or in a new issue.

**Comment by sandwichsudo** [comment]
For anyone that comes after - I saw this in 7.4.0 [sandbox](https://codesandbox.io/s/5nz09l2yp) but it's fixed in 7.4.2.
